File: src/OpenAiUtils.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

module OpenAiUtils
  ( makeChatRequest
  , ChatMessage(..)
  , ChatRequest(..)
  , ChatChoice(..)
  , ChatResponse(..)
  ) where

import Network.HTTP.Client
import Network.HTTP.Client.TLS
import Data.Aeson
import Data.Text (Text)
import qualified Data.Text as T
import Data.Text.Encoding (encodeUtf8)
import qualified Data.ByteString.Lazy as B
import Control.Exception (SomeException, try)
import System.Environment (getEnv)
import GHC.Generics

data ChatRequest = ChatRequest
  { model :: Text
  , messages :: [ChatMessage]
  , response_format :: Value
  , temperature :: Double
  , max_tokens :: Int
  } deriving (Show, Generic)

instance ToJSON ChatRequest

data ChatMessage = ChatMessage
  { role :: Text
  , content :: Text
  } deriving (Show, Generic)

instance ToJSON ChatMessage
instance FromJSON ChatMessage


newtype ChatResponse
  = ChatResponse {choices :: [ChatChoice]}
  deriving (Show, Generic)

instance FromJSON ChatResponse

newtype ChatChoice
  = ChatChoice {message :: ChatMessage}
  deriving (Show, Generic)

instance FromJSON ChatChoice


makeChatRequest ::  Value -> [ChatMessage] -> IO (Either String B.ByteString)
makeChatRequest schema messages = do

  apiKey <- getEnv "OPENAI_API_KEY"
  let managerSettings = tlsManagerSettings { managerResponseTimeout = responseTimeoutMicro (60 * 1000000) } 
  manager <- newManager managerSettings
  let url = "https://api.openai.com/v1/chat/completions"
      requestBody = encode ChatRequest
        { model = "gpt-4o"
        , messages = messages
        , response_format = schema
        , temperature = 0.0
        , max_tokens = 3000
        }
  initialRequest <- parseRequest $ T.unpack url
  let request = initialRequest
        { method = "POST"
        , requestHeaders =
            [ ("Content-Type", "application/json")
            , ("Authorization", "Bearer " <> encodeUtf8 (T.pack apiKey))
            ]
        , requestBody = RequestBodyLBS requestBody
        }
  response <- try (httpLbs request manager) :: IO (Either SomeException (Response B.ByteString))
  case response of
    Left err -> return $ Left (show err)
    Right res -> return $ Right (responseBody res)

File: fullstack/backend/app/open_ai_utils.py
import logging
import os
from typing import TypeVar

import openai
from pydantic import BaseModel

logger = logging.getLogger(__name__)


class ChatMessage(BaseModel):
    role: str
    content: str


type Prompt = list[ChatMessage]

T = TypeVar("T", bound=BaseModel)


def make_chat_request(model: type[T], messages: list[ChatMessage]) -> T | None:
    """Send a chat request to OpenAI and return the response as a Pydantic model."""

    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        raise ValueError("Environment variable 'OPENAI_API_KEY' not set.")

    client = openai.OpenAI(api_key=api_key)

    # try:
    response = client.beta.chat.completions.parse(
        model="gpt-4o-mini",
        messages=[msg.model_dump() for msg in messages],  # type: ignore[misc]
        response_format=model,
        temperature=0.0,
        max_tokens=9000,
    )

    if not response.choices:
        raise ValueError("No response choices received from OpenAI.")

    val = response.choices[0].message.parsed
    return val

    # except openai.OpenAIError as e:
    #    print(f"OpenAI API error: {e}")
    #    return None
